Optimizer infrastructure

wjakob opened the pull request:
This commit adds standard gradient-based optimizers to the Dr.Jit codebase:

- ``drjit.opt.SGD`` (optionally with momentum)
- ``drjit.opt.RMSProp``
- ``drjit.opt.Adam``

It also adds tests that compare their output to reference data obtained from PyTorch.

Speierers commented on drjit/opt.py:
```diff
@@ -0,0 +1,1192 @@
+            self._reset(key, value)
+
+    def __len__(self) -> int:
+        """Return the number of number of registered parameters."""
```
typo
```suggestion
        """Return the number of registered parameters."""
```

wjakob replied:
Thanks

wjakob commented:
I did another pass over the Optimizer API and incorporated the following changes.

- Most of the shared functionality has now been moved to `Optimizer.step()`.
- The optimizer-specific update rules are in a methods of the form `OptimizerSubclass._step`. This is quite a bit less code compared to before.
- The `mask_update` feature is part of `Optimizer`, which means that it is now also available in `SGD`, `RMSProp`, etc.
- We now run `dr.make_opaque()` on anything that is being put into the optimizer.
- Fixed optimization of tensors and added a test.

njroussel commented on drjit/opt.py:
```diff
@@ -0,0 +1,1199 @@
+                # Update the optimizer state and schedule it for evaluation
+                new_state = new_value, lr, new_extra
+
+                dr.schedule(new_state)
```
If `eval == False`, I was almost willing to suggest that we don't even `dr.schedule` the new state, so that users have the full flexibility to decide what gets evaluated.

But it seems like too much of a footgun, users can/should just remove the parameters that they don't want to compute from the optimizer rather than relying on it to not be evaluated.

njroussel reviewed (approved):
:rocket:  all good
